File: Dai-Hentai/ViewControllers/TabBarController/TabBarController.m
//
//  TabBarController.m
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2018/1/9.
//  Copyright © 2018年 DaidoujiChen. All rights reserved.
//

#import "TabBarController.h"
#import "GalleryViewController.h"
#import "SettingViewController.h"

@interface TabBarController () <UITabBarControllerDelegate>

@end

@implementation TabBarController

#pragma mark - UITabBarControllerDelegate

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController {
    for (UIViewController *controller in tabBarController.viewControllers) {
        if (![controller isKindOfClass:[UINavigationController class]]) {
            continue;
        }
        
        UINavigationController *navigation = (UINavigationController *)controller;
        if ([navigation.topViewController isKindOfClass:[GalleryViewController class]]) {
            GalleryViewController *gallery = (GalleryViewController *)navigation.topViewController;
            [gallery.navigationController popViewControllerAnimated:NO];
        }
        else if ([navigation.topViewController isKindOfClass:[SettingViewController class]]) {
            SettingViewController *setting = (SettingViewController *)navigation.topViewController;
            [DBUserPreference setInfo:setting.info];
        }
    }
}

#pragma mark - Life Cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.delegate = self;
    
    __block NSMutableArray *titles = [NSMutableArray arrayWithObjects:@"列表", @"歷史", @"下載", @"設定", nil];
    [self.tabBar.subviews enumerateObjectsUsingBlock: ^(UIView *v, NSUInteger tabBarIndex, BOOL *tabBarStop) {
        if ([v isKindOfClass:NSClassFromString(@"UITabBarButton")] && v.subviews.count == 2) {
            [v.subviews enumerateObjectsUsingBlock: ^(UIView *buttonSubView, NSUInteger buttonIndex, BOOL *buttonStop) {
                if ([buttonSubView isKindOfClass:NSClassFromString(@"UITabBarButtonLabel")]) {
                    [buttonSubView performSelector:@selector(setText:) withObject:titles.firstObject];
                    [buttonSubView sizeToFit];
                    [titles removeObjectAtIndex:0];
                    *buttonStop = YES;
                }
            }];
        }
    }];
}

@end

File: Dai-Hentai/ViewControllers/TabBarController/TabBarController.swift
//
//  TabBarController.swift
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2018/11/27.
//  Copyright © 2018 DaidoujiChen. All rights reserved.
//

import Foundation
import UIKit

class TabBarController: UITabBarController {
    
    weak var lastSelectedItem: UITabBarItem? = nil
    
    // MARK: - Life Cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.delegate = self
        var titles = [ "列表", "歷史", "下載", "設定" ]
        let selector = Selector(("setText:"))
        for view in tabBar.subviews {
            if
                let tabBarButton = NSClassFromString("UITabBarButton"),
                view.isKind(of: tabBarButton),
                view.subviews.count == 2 {
                
                for subview in view.subviews {
                    if
                        let tabBarButtonLabel = NSClassFromString("UITabBarButtonLabel"),
                        subview.isKind(of: tabBarButtonLabel),
                        subview.responds(to: selector) {
                        
                        subview.perform(selector, with: titles.first ?? "")
                        subview.sizeToFit()
                        titles.remove(at: 0)
                        break
                    }
                }
            }
        }
    }
    
}

// MARK: - UITabBarControllerDelegate
extension TabBarController: UITabBarControllerDelegate {
    
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        if
            let navigationController = viewController as? UINavigationController,
            let settingViewController = navigationController.topViewController as? SettingViewController {
            DBUserPreference.setInfo(settingViewController.info)
        }
    }
    
    override func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
        if lastSelectedItem == item {
            NotificationCenter.default.post(name: NSNotification.Name(rawValue: "ListScrollToBottom"), object: nil)
        }
        lastSelectedItem = item
    }
    
}
